### ISSUE TITLE
Adding non-serialisable metadata to toState

### ISSUE THREAD

**Issue by satazor** [router5/router5#135]
This is a follow up of https://github.com/router5/router5/issues/60#issuecomment-233296870

@troch I'm hitting the issue that @iotch mentioned.. I would like to attach non-serializable stuff into `toState`: 

```js
// getComponentMiddleware.js
export default (routes) => (router) => (toState, fromState) => {
    const { toActivate } = transitionPath(toState, fromState);

    // Get the routes involved in the transition
    const routesToActivate = toActivate.map((segment) => routes.find((route) => route.name === segment));

    // Build a map in which keys are the route names and values that resolve to their components
    const promisesMap = fromPairs(routesToActivate.map((route) => [route.name, route.component || route.getComponent()]));

    return pProps(promisesMap)
    .then((components) => ({ ...toState, components }));
};
```

`DOMException: Failed to execute 'pushState' on 'History': An object could not be cloned.`

So here are a few questions:
- Do we really have to persist the state to the history api? `browser.getState` is only used [here](https://github.com/router5/router5/blob/master/modules/plugins/browser/index.js#L103). Isn't it more correct to compare the URLs instead of comparing states? In my opinion, the browser plugin is just a way to set and react to URL changes, it shouldn't be storing nor handling states. Also, this would remove the inconsistency that exists now between using history API or hashbangs.
- If you would like to keep it like this, could we have field that would not be included in the history state? This would allow us to add big data structures (big structures generate QUOTA errors) or unserializable data to be added to `toState`.

Thanks in advance.

**Comment by satazor** [comment]
@troch I'm really trying to use router5 and it looks like a great router to me, both conceptually and in terms of execution.

This middleware is just one of the ideas that I have to improve `react-router5`. Perhaps could we discuss this in gitter or something similar?

**Comment by motet-a** [comment]
Personally, I find really interesting to persist the state in the browser history, in order to retreive “hidden” parameters when the user navigates with back and forward buttons.

However — tell me if I’m wrong — I think you can actually do what you want if you inject a mocked browser in the plugin factory, [at the second argument](https://github.com/router5/router5/blob/master/modules/plugins/browser/index.js#L14):

```js
import defaultBrowser from 'router5/plugins/browser/browser'

let savedState
                         
const myCustomBrowser = {
    ...defaultBrowser,
    pushState(state, title, path) {
        window.history.pushState({}, title, path)
        savedState = state
    },

    replaceState() {
        // Similar hack
    },

    getState() {
        return savedState
    },
    
    // Unfortunately, you’ll probably also need to hack
    // `addPopstateListener` and `removePopstateListener`.
}

router.usePlugin(browserPlugin({}, myCustomBrowser))
```

There is [something similar here](https://github.com/router5/router5/blob/master/tests/plugins/browser.js).

In anyway, it’s probably too cumbersome and hackish and I’d be glad to know better ways to achieve it.

**Comment by satazor** [comment]
@motet-a you are right, we are able to do that. But I would like to avoid that and only use it as last resort.
 
One of the strong reasons to not store the state is that, right now, the browser plugin behave differently when using hash or not. But I'm fine if you want to keep the current functionality but I think we should be able to attach metadata into something like state.locals, quite similar to res.locals from express.

**Comment by motet-a** [comment]
Yes. I think the best is to allow the user to customize entirely the way the state is stored with simple options. I think something in the spirit of `react-redux` could be suitable:

```js
let savedBigData

function mapRouterStateToBrowserState({ bigData, ...state }) {
    savedBigData = bigData
    return state // Will be saved in window.history.state (bigData has been extracted)
}

function mapBrowserStateToRouterState(browserState) {
    return {
        ...browserState, 
        bigData: savedBigData,
    }
}

router.usePlugin(
    browserPlugin({
        mapRouterStateToBrowserState,
        mapBrowserStateToRouterState,
    })
)
```

`mapRouterStateToBrowserState` and `mapBrowserStateToRouterState` can both default to an identity function, `s => s`. This is not breaking changes.

**Comment by troch** [comment]
@satazor the state is added to the history so when going back and forward, paths don't have to be matched and states are already built.

I don't believe using hash or not makes a difference: in both cases the history API is used. What difference do you see?

**Comment by troch** [comment]
@satazor this has been fixed with v4.7.1: only base state properties (`meta`, `name`, `path` and `params`) are pushed to history.
